### ISSUE TITLE
[bug] version is not bumped when scope is specified

### ISSUE THREAD

**Issue by m4t22** [mooltiverse/nyx#115]
Hello,

I've been using the plugin with version `1.2.1` (latest at the time of writing) and it is not bumping version if there is scope specified.

Examples:
- `feat: cool feature` - will trigger bump version
- `feat(JIRA-TICKET-ID): cool feature` - will not trigger bump version 

My nyx config from `build.gradle`
```
nyx {
  preset = 'simple'
  releaseLenient = true
  releasePrefix = 'v'
  initialVersion = '0.1.0'
  releaseTypes {
    enabled = [ 'mainline' ]
    items {
      mainline {
        gitTag = 'true'
      }
    }
  }
}
```

**Comment by m4t22** [comment]
I've used `commitMessageConventions` configuration block and updated `expression` regex to the following one:
```
commitMessageConventions {
    enabled = [ 'conventionalCommits' ]
    items {
      conventionalCommits {
        expression = '(?m)^(?<type>[a-zA-Z0-9_]+)(!)?(\\((?<scope>[A-Z0-9\\-]+)\\))?:( (?<title>.+))$(?s).*'
        bumpExpressions {
          major = '(?s)(?m)^[a-zA-Z0-9_]+(!|.*^(BREAKING( |-)CHANGE: )).*'
          minor = '(?s)(?m)^feat(?!!|.*^(BREAKING( |-)CHANGE: )).*'
          patch = '(?s)(?m)^fix(?!!|.*^(BREAKING( |-)CHANGE: )).*'
        }
      }
    }
  }
```
but then I got tags doubled for a single commit after single `./gradlew release` command execution

**Comment by m4t22** [comment]
Hello @flelli, 

I am sorry, I was convinced you can put pretty everything into `scope`.
As for double tags this is what happens when I execute `./gradlew release`.
Initial output looks good:

<img width="353" alt="Screenshot 2022-10-19 at 09 13 47" src="https://user-images.githubusercontent.com/44002313/196621960-0339bb94-02a5-4577-a3cd-46573be24e3d.png">

But once you run `git log`, there are 2 tags for the same commit, that was previously untagged:

<img width="441" alt="Screenshot 2022-10-19 at 09 14 37" src="https://user-images.githubusercontent.com/44002313/196622110-846befff-761c-4f85-912e-3ac843aa339d.png">

and this is the sanitized state file content: https://pastebin.com/CsKv69D6

**Comment by flelli** [comment]
Hi @m4t22 ,

this actually looks like an unexpected behavior.

A few questions to help me out triaging:

- have the two tags `v0.6.0` and `v0.5.0` been applied during the same run or 2 separate runs?
- if 2 separate runs, did the configuration change by any means?
- is this a public repo I can clone to reproduce the error?

One thing that caught my attention in the state file is that commit f4f26fd0f1c9f9d05900aec229a9a2954becb2c4 appears as tagged with `v0.5.0` in some contexts (previousVersionCommit) but not tagged at att in others (releaseScope, significantCommits, initialCommit, finalCommit). This may be relevant in triaging the defect.

**Comment by m4t22** [comment]
Hello, 

- it has been single run - this is why it is so strange
- it's private, but I can share gradle.build config:

```
nyx {
  releasePrefix = 'v'
  releaseTypes {
    enabled = [ 'mainline' ]
    items {
      mainline {
        gitTag = 'true'
      }
    }
  }

  commitMessageConventions {
    enabled = [ 'conventionalCommits' ]
    items {
      conventionalCommits {
        expression = '(?m)^(?<type>[a-zA-Z0-9_]+)(!)?(\\((?<scope>[A-Z0-9\\-]+)\\))?:( (?<title>.+))$(?s).*'
        bumpExpressions {
          major = '(?s)(?m)^[a-zA-Z0-9_]+(!|.*^(BREAKING( |-)CHANGE: )).*'
          minor = '(?s)(?m)^feat(?!!|.*^(BREAKING( |-)CHANGE: )).*'
          patch = '(?s)(?m)^fix(?!!|.*^(BREAKING( |-)CHANGE: )).*'
        }
      }
    }
  }
}
```

**Comment by flelli** [comment]
I tried to reproduce the behavior but with no success (meaning it worked as expected).

Here's what I've done so maybe you can spot if there are relevant changes:

```
# create a brand new Git repository and cd into it
git init nyx-triage-115
cd nyx-triage-115
# create the build.gradle file (content below) and do the first commit, tagged as `v0.4.0`
git add .
git commit -a -m "feat(WX-6206): Multibranch build pipelines DSL (#13)\n"
git tag v0.4.0
# now the only commit in the repo looks like commit a45a488c986c43c030f2f3c670c4fae04524d62f (tagged as v0.4.0) from the triage case
git log
commit e04c8d247967ef891a61be2ec7de7b504227abc1 (HEAD -> master, tag: v0.4.0)
[...]

    feat(WX-6206): Multibranch build pipelines DSL (#13)\n
# add a README.txt and the .gitignore file to ignore the state file output and the .gradle folder
# then create a new commit that will be tagged by Nyx
git add .
git commit -a -m "feat(WX-0): test commit\n"
# now the git repository looks like the triage case before running Nyx, with the last commit that looks like commit f4f26fd0f1c9f9d05900aec229a9a2954becb2c4 from the case before tagging
git log
commit 0c520739059673d9c0cedde6cb5f501b4282c7a6 (HEAD -> master)
[...]

    feat(WX-0): test commit\n

commit e04c8d247967ef891a61be2ec7de7b504227abc1 (tag: v0.4.0)
[...]

    feat(WX-6206): Multibranch build pipelines DSL (#13)\n
# run Nyx by means of "gradle release"
gradle release --debug
git log
commit 0c520739059673d9c0cedde6cb5f501b4282c7a6 (HEAD -> master, tag: v0.5.0)
[...]

    feat(WX-0): test commit\n

commit e04c8d247967ef891a61be2ec7de7b504227abc1 (tag: v0.4.0)
[...]

    feat(WX-6206): Multibranch build pipelines DSL (#13)\n
```

This way the defect is not reproduced as only tag `v0.5.0` is applied, while the submitted case had a double tag `v0.5.0` and `v0.6.0`.

The content of the `build.gradle` file is (I only added the `stateFile = 'nyx-state.json'` starting from your content):

```
plugins {
  id "com.mooltiverse.oss.nyx" version "1.2.1"
}

nyx {
  releasePrefix = 'v'
  releaseTypes {
    enabled = [ 'mainline' ]
    items {
      mainline {
        gitTag = 'true'
      }
    }
  }

  commitMessageConventions {
    enabled = [ 'conventionalCommits' ]
    items {
      conventionalCommits {
        expression = '(?m)^(?<type>[a-zA-Z0-9_]+)(!)?(\\((?<scope>[A-Z0-9\\-]+)\\))?:( (?<title>.+))$(?s).*'
        bumpExpressions {
          major = '(?s)(?m)^[a-zA-Z0-9_]+(!|.*^(BREAKING( |-)CHANGE: )).*'
          minor = '(?s)(?m)^feat(?!!|.*^(BREAKING( |-)CHANGE: )).*'
          patch = '(?s)(?m)^fix(?!!|.*^(BREAKING( |-)CHANGE: )).*'
        }
      }
    }
  }
  stateFile = 'nyx-state.json'
}
```

There must be something different between my script here and your repo. Can you help me in reproducing the scenario?

**Comment by m4t22** [comment]
How about gradle version? I have: `Gradle 7.5.1`

**Comment by flelli** [comment]
I first tested with 7.4 but running the test again with 7.5.1 yields to the same outcome.

Any change you can run the same test above and see what happens on your side?

**Comment by m4t22** [comment]
I was able to reproduce it on this repo: https://github.com/m4t22/gradle-test 
I have: 
- run `gradle init` 
  - set up project for groovy
- edited `build.gradle` to contain the nyx configuration
- created init commit
- run `./gradlew release` - got 2 tags for single run

**Comment by flelli** [comment]
I cloned your repo and executed `./gradlew release` but no new tag was applied.
This is intriguing... :)

Do you mind running the same with `./gradlew release --debug` and sending the debug output, alog with the outputh from a `git log` before and after running Gradle?

**Comment by m4t22** [comment]
as I saw: `Debug level logging will leak security sensitive information!`, I've removed the tags and you can reproduce it on your end on single untagged commit.

**Comment by flelli** [comment]
I know it gives you that message but if you run it on the test project there should be nothing sensitive in there.

In case it helps I'm only interested in messages containing the `com.mooltiverse.oss.nyx` string so you can filter out all the others and check there's no sensitive information in what is left.

I understand your concern and I don't want to push you but it seems all the tests I run on my end are producing the right outcomes so I can't reproduce the error.

This seems to be a very weird behavior and as far as I can imagine it should definitely not happen as the code is structured in a way that the tagging function is traversed only once.

Lacking of a reproducible case I can only wonder if there's any external agent affecting tags like git hooks or other tools that kick in in the background by some other means.

**Comment by m4t22** [comment]
git log before:
```
commit db888a5731a14b618494b2c9c878372797b93228 (HEAD -> main, origin/main)
Author: Mateusz <gmail.com>
Date:   Wed Oct 19 15:01:03 2022 +0200

    feat: gradle init
```

filtered debug log: [debug.log](https://github.com/mooltiverse/nyx/files/9830307/debug.log)

git log after:
```
commit db888a5731a14b618494b2c9c878372797b93228 (HEAD -> main, tag: v0.3.0, tag: v0.2.0, origin/main)
Author: Mateusz <gmail.com>
Date:   Wed Oct 19 15:01:03 2022 +0200

    feat: gradle init
```

**Comment by flelli** [comment]
``I was finally able to reproduce the bug on a Mac only, using [your test project](https://github.com/m4t22/gradle-test), but at least I have something.

My findings so far are:

1. it happens on the Mac only (which is pretty weird)
2. it has nothing to deat with Gradle parallelism or using Nyx as a project/settings plugin (I had a doubt on this)
3. removing the `commitMessageConventions` block from the configuration works correctly (although it doesn't match all messages from the convention you are setting up)

I'm going to investigate this deeper.
In the meanwhile, could you please confirm that with a configuration like

```
nyx {
  releasePrefix = 'v'
  releaseTypes {
    enabled = [ 'mainline' ]
    items {
      mainline {
        gitTag = 'true'
      }
    }
  }
}
```

it no longer duplicates tags?

**Comment by m4t22** [comment]
I can confirm, I was changing configuration back and forth, only after adding `commitMessageConventions` block issue appeared on my side.

**Comment by flelli** [comment]
Good news, the behavior is consistent on all platforms and does not only affect the Mac.

I was able to spot the bug and it has to deal with some internal up-tp-date checks in case of _dirty_ workspaces (containing uncomiitted changes).

I'm going to close this issue and open a new bug linked to this one.

In the meanwhile a simple workaround that you may use is storing (and committing) a `.gitignore` file in the repository. The one I used in the test repository you shared has these 2 lines:

```
/.gradle/
nyx-state.json
```

Add any other contents that you don't need to commit to the file.

Once also the `.gitignore` is committed you can see the repository status is _clean_ by running `git status`. This prevents the tool from running into the faulty condition.

**Comment by m4t22** [comment]
Hello @flelli,

I can confirm that it works perfectly! Thank you for handling the issue so quickly!

**Comment by flelli** [comment]
That was just a temporary workaround while I'm still working on the fix (#116), but I'm glad you got it working for the meanwhile.
I'll keep you posted when the fix is ready.

**Comment by flelli** [comment]
The bug has been fixed in release [1.2.2](https://github.com/mooltiverse/nyx/releases/tag/1.2.2).

Thanks @m4t22 for finding the bug and contributing the info.

**Comment by m4t22** [comment]
Hello, thank you @flelli, I wish all github issues in other projects have been solved that quickly! :)
